Implement logic to divide bindings into disc images

helhotti opened the pull request.

aajarven commented on harvester/utils.py:
```diff
@@ -119,11 +113,42 @@ def calculate_batch_size(col_size):
+    Split a list of bindings into images, each image containing no more than
+    max_bindings_per_image bindings.
+    """
+    images = []
+    for i in range(10):
+        prefix = shared_prefix + str(i)
+        prefixed_bindings = bindings_with_prefix(bindings, prefix)
+        if len(prefixed_bindings) <= max_bindings_per_image:
+            images.append({"prefix": prefix, "bindings": prefixed_bindings})
+        else:
+            images.extend(
+                assign_bindings_to_images(
+                    prefixed_bindings, max_bindings_per_image, prefix
+                )
+            )
+    return images
```
re: RC discussion about not splitting small collections at all. We can make them go into a single image by checking whether an image is small enough not after it has been split but before, something like this:
```
    if len(bindings) <= max_bindings_per_image:
        return [{"prefix": shared_prefix, "bindings": bindings}]

    images = []
    for i in range(10):
        prefix = shared_prefix + str(i)
        prefixed_bindings = bindings_with_prefix(bindings, prefix)
        results.extend(
            assign_bindings_to_images(prefixed_bindings, max_bindings_per_image, prefix)
            )
    return images
```

aajarven reviewed (changes requested):
We still need tests (trying to sell me code without tests when the code is by me will not fly, I don't trust the author) and I added a suggestion for making it possible to not always split the smallest collections, but otherwise things are looking good.

In the long term, we need to think about the output format of the splitting function so that it's easy to use for the caller. Might be that the dict is fine though, but it might also turn out to be annoying. We will also want to synchronize `assign_bindings_to_images` and `split_into_download_batches` so that one batch doesn't span two or more images, but that's also something we can implement later when we get to actually creating images.

helhotti commented:
> We still need tests (trying to sell me code without tests when the code is by me will not fly, I don't trust the author) and I added a suggestion for making it possible to not always split the smallest collections, but otherwise things are looking good.

Tests are now added and the splitting skipping is implemented as you suggested

> In the long term, we need to think about the output format of the splitting function so that it's easy to use for the caller. Might be that the dict is fine though, but it might also turn out to be annoying. 

I would leave the dicts for now, for the simple reason that I will not have time to thoroughly think about this during this sprint.

> We will also want to synchronize `assign_bindings_to_images` and `split_into_download_batches` so that one batch doesn't span two or more images, but that's also something we can implement later when we get to actually creating images.

My preliminary idea for this is that dividing the bindings to disk images is the first step after a collection of binding ids is acquired, and splitting into download batches is done in the next step. This way, the whole download procedure can be done for one image per TaskGroup (or similar), so no mix ups should happen.

helhotti commented on harvester/utils.py:
```diff
@@ -119,11 +113,39 @@ def calculate_batch_size(col_size):
+        return [{"prefix": shared_prefix, "bindings": bindings}]
+    images = []
+    for i in range(10):
+        prefix = shared_prefix + str(i)
```
The other option would be that we create empty images, right? I guess in the future if we encounter new bindings with a prefix that does not match to an existing image, we create a new image with the same logic as the existing ones.

Traubert replied:
The logic will be a bit complicated: if you see that 123/ doesn't exist, you can't assume that either 12/, 1/ or / is the correct image, because there might exist any of 121/, 122/, 124/, etc.

Buut I guess that's not such a big deal. If @aajarven also says no problem, this looks good to me!

aajarven replied:
This solution does indeed return image 0 (with zero bindings inside), as well as other empty images with siblings that do contain bindings. For example dividing collection collection-361 with max 200 bindings/image (nb: totally toy example, in reality this is such a small collection that we would likely have it in one image) produces the following `prefix: n_bindings` pairs:
```
0:	0
1:	139
20:	105
21:	82
22:	19
23:	0
24:	0
25:	0
26:	0
27:	0
28:	0
29:	0
3:	0
4:	0
5:	0
6:	0
7:	0
8:	0
9:	0
```
What we do with these empty bindings is then left as an exercise for the programmer when we get to KP-7020. In many ways it would simplify life for both us and our users if we created the empty images, but in terms of mount points that would be somewhat wasteful. At least the zero-image can and should be omitted, but whether it should be done here or by the caller depends on what the caller will end up looking like.

Not creating the empty images does indeed make the incremental addition to pre-existing images more difficult, as we need to check what is the longest prefix length at which the newly-added binding still has siblings. Slightly more complex and a bit more error-prone, but I don't see why that would not be doable.

But all in all, unless we already know what we want, I'd leave the tweaking for later, as doing anything too specific now runs a high risk of optimizing for something that will not be our actual solution.

helhotti replied:
Good points from both of you! We need to keep all this in mind in KP-7020, but I would agree that since we don't have a clear idea about what want, we should leave this for the next ticket and preferably have some discussions together thinking of the pros and cons of different approaches.

aajarven reviewed (approved):
I say :+1:.

If we wanted to be really pedantic, the tests could test the functionality more thoroughly (we only create "leaf" images, e.g. not both 123 and 1234; we return the siblings and the fact that they are empty, the bindings in an image really have the correct prefix, different branches of the tree can have different heights...) in separate test functions (for easy readability of the output when something eventually fails). But this is all just haihattelu and unit test best practice propaganda, no need to make changes now.
